reutils/multimatch.go

```diff
@@ -0,0 +1,55 @@
+package reutils
+
+import (
+	"regexp"
+	"sync"
+)
+
+// MultiMatch returns a map of matching elements from a list of regular expressions (returning the first matching element)
+func MultiMatch(s string, expressions ...*regexp.Regexp) (map[string]string, int) {
+	for i, re := range expressions {
+		if matches := re.FindStringSubmatch(s); len(matches) != 0 {
+			results := make(map[string]string, len(matches))
+			results[""] = matches[0]
+			for i, key := range re.SubexpNames() {
+				if key != "" {
+					results[key] = matches[i]
+				}
+			}
+			return results, i
+		}
+	}
+	return nil, -1
+}
+
+// NewRegexGroup cache compiled regex to avoid multiple interpretation of the same regex
+func NewRegexGroup(key string, definitions ...string) (result []*regexp.Regexp, err error) {
+	result = make([]*regexp.Regexp, len(definitions))
+	for i := range definitions {
+		regex, err := regexp.Compile(definitions[i])
+		if err != nil {
+			return nil, err
+		}
+		result[i] = regex
+	}
+
+	cachedRegex.Store(key, result)
+	return
+}
+
+// GetRegexGroup tries to retreive a regular expression group previously created by NewRegexGroup.
+func GetRegexGroup(key string) (result []*regexp.Regexp) {
```

Comment:
Do we need this function?
Can't we just have a `Get` function which creates the key if it doesn't exist, and fetches from cache otherwise?
Seems error prone to have a function that needs to have items added to it before being called

Reply:
The problem with the previous version was that it ignored the arguments if the key existed. But it was error prone in the case where the arguments to NewRegexGroup were changing. Then, the function returned the cache and I had no way to overwrite the values.